src/jit/flowgraph.cpp

```diff
@@ -22471,3 +22471,240 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+
+        // If there are more callfinallys than continuations, some of the
+        // callfinallys must share a continuation, and we can merge them.
+        const bool tryMerge = callFinallyCount > continuationMap.GetCount();
+
+        if (!tryMerge)
+        {
+            JITDUMP("EH#%i does not have any mergeable callfinallys\n", XTnum);
+            continue;
+        }
+
+        canMerge = true;
+
+        // Walk the try and callfinally regions, looking for blocks
+        // that jump to a callfinally that invokes this try's finally,
+        // and make sure they all jump to the appropriate canonical
+        // callfinally.
+        BasicBlock* const endTryBlock = HBtab->ebdTryLast->bbNext;
+        for (BasicBlock* currentBlock = HBtab->ebdTryBeg; currentBlock != endTryBlock;
+             currentBlock             = currentBlock->bbNext)
+        {
+            bool merged = fgRetargetBranchesToCanonicalCallFinally(currentBlock, beginHandlerBlock, continuationMap);
+            didMerge    = didMerge || merged;
+        }
+
+        for (BasicBlock* currentBlock = firstCallFinallyRangeBlock; currentBlock != endCallFinallyRangeBlock;
+             currentBlock             = currentBlock->bbNext)
+        {
+            bool merged = fgRetargetBranchesToCanonicalCallFinally(currentBlock, beginHandlerBlock, continuationMap);
+            didMerge    = didMerge || merged;
+        }
+    }
+
+#if DEBUG
+
+    if (!canMerge)
+    {
+        JITDUMP("Method had try-finallys, but did not have any mergeable finally chains.\n");
+        return;
+    }
+
+    assert(didMerge);
+    JITDUMP("Method had try-finallys and some callfinally merges were performed.\n");
+
+    if (verbose)
+    {
+        printf("\n*************** After fgMergeFinallyChains()\n");
+        fgDispBasicBlocks();
+        fgDispHandlerTab();
+        printf("\n");
+    }
+
+#endif // DEBUG
+}
+
+//------------------------------------------------------------------------
+// fgRetargetBranchesToCanonicalCallFinally: find non-canonical callfinally
+// invocations and make them canonical.
+//
+// Arguments:
+//     block -- block to examine for call finally invocation
+//     handler -- start of the finally region for the try
+//     continuationMap -- map giving the canonical callfinally for
+//        each continuation
+//
+// Returns:
+//     true iff the block's branch was retargeted.
+
+bool Compiler::fgRetargetBranchesToCanonicalCallFinally(BasicBlock*      block,
+                                                        BasicBlock*      handler,
+                                                        BlockToBlockMap& continuationMap)
+{
+    // We expect callfinallys to be invoked by a BBJ_ALWAYS at this
+    // stage in compilation.
+    if (block->bbJumpKind != BBJ_ALWAYS)
+    {
+        // Possible paranoia assert here -- no flow successor of
+        // this block should be a callfinally for this try.
+        return false;
+    }
```

Comment:
I think you are saying that on ARM32 the address of the continuation is passed to the finally as the resume point, instead of the address of the always block. But the successor of the always block is the method the jit uses to identify the continuation, so changing the successor like this seems on the surface like it ought to work. And since we're only swapping one continuation for another the new successor should have the finally target bit set.

I still suspect something funny may be going on with the finally target bit, because chain merging will create more cases where a such a block has multiple predecessors. But no hard evidence yet.